### PR TITLE
Add symbols from Orca

### LINKED ISSUE #11105: Integrating Orca's symbol translations?
Hello,

On the Linux side, we are working on making the symbols processing more coherent, basically basing on the NVDA tables so we can have shared open-source effort there on fixes and translations.

There are a few symbols that Orca currently translates which are not yet in the NVDA tables, such as □, ○, ◆, ✗, ¼, ½, ¾, ⁰, ⁴, etc. with a lot of available language translations. I can easily work on providing a patch that simply adds them all to `source/locale/*/symbols.dic`, would that be enough for NVDA, or is there more to it? (e.g. to integrate with the translation process)

Samuel

### PR DESCRIPTION
**Pull request by sthibaul**
# Link to issue number:

fixes #11105

### Summary of the issue:

Some symbols have been seen useful in Orca and are missing in NVDA

### Description of how this pull request fixes the issue:

This adds them to source/locale/en/symbols.dic

### Testing performed:

The resulting file processes fine.

### Known issues with pull request:

This adds a series of symbols to translate. Orca translations can be used as a source if translators wish. I will attach a tarball containing them.

### Change log entry:

*New features*
`Adds more symbols. (#11105)`

### CODE REVIEW

**Comment by sthibaul** [comment]
Here are the Orca translations, in case translators wish to reuse them:
[orca.zip](https://github.com/nvaccess/nvda/files/4578463/orca.zip)

**Comment by codeofdusk** [comment]
NVDA seems to be able to read these at least when reading by character...

**Comment by Adriani90** [comment]
If they are not in NVDA yet, tese symbols need also to be added to the symbols and interpunctuation dialog. Actually there are a lot of symbols.
Adding new symbols to symbols.dic makes them translatable and defines the default value.
I don't know exactly if the symbols are added automatically to the dialog once they are added to symbols.dic, but I think this might need to be done manually unfortunately.

**Comment by josephsl** [comment]
Hi, punctuation/symbol pronunciation dialog will pick up symbols from language-specific symbols.dic, which ultimately depends on willingness from translators to translate new symbols once they show up on translations workflow. Thanks.

**Comment by CyrilleB79** [comment]
Just a question: is there some guideline or some common usage to decide which symbol is added in symbol.dic and which should not? Is there a big performance penalty to add many symbols?
I have raised theses questions in #11105 with some more explanations and example. Please let's answer and comment there.

**Comment by k-kolev1985** [comment]
> 
> 
> Here are the Orca translations, in case translators wish to reuse them:
> [orca.zip](https://github.com/nvaccess/nvda/files/4578463/orca.zip)

May I ask why the engb (english) one is so small/short? There are only a few entries in it. The russian one is also not complete. I want those two (at least the english one) for reference while discussing them with my colleagues for the bulgarian translation. Thanks in advance!

**Comment by sthibaul** [comment]
> May I ask why the engb (english) one is so small/short? 

It's just because my script ignores the strings identical with the "en" version

**Comment by josephsl** [comment]
Hi, in other words, what translators would have seen in symbols changes folder for a given revision minus diff signs. Thanks.

**Comment by Adriani90** [comment]
@josephsl wrote:
> Hi, punctuation/symbol pronunciation dialog will pick up symbols from language-specific symbols.dic, which ultimately depends on willingness from translators to translate new symbols once they show up on translations workflow. Thanks.

But in the punctuation and symbol pronounciation dialog, there are more than 3.000 symbols. Where are they fetched from? In the symbols.dic are only about 200 lines. If they are fetched from Windows itself, I think most of them should be already translated (i.e. emojis). What is exactly the purpose of the symbols.dic then?

**Comment by josephsl** [comment]
Hi, some of them come from symbols.dic, others are emoji characters coming from CLDR database. Thanks.

File: source/locale/en/symbols.dic [hunk 1]
```diff
@@ -172,6 +230,9 @@ _	line	most
 ⊇	superset of or equal to	none
 
 # Vulgur Fractions U+2150 to U+215E
+¼	one fourth	none
+½	one half	none
+¾	three fourths	none
```
Comment (feerrenrut) [inline source/locale/en/symbols.dic:235]: I believe "one quarter" and "three quarters" is much more popular than the also acceptable "fourth" version. I propose this be changed. 
https://english.stackexchange.com/questions/103188/three-quarters-vs-three-fourths

**Review by feerrenrut — CHANGES_REQUESTED** [review]

**Comment by CyrilleB79** [comment]
I am copy/pasting my comment from #11105 since it seems that the discussion is taking place here.

> Just a question: is there some guideline or some common usage to decide which symbol is added in symbol.dic and which should not? Is there a big performance penalty to add many symbols?
> I raise these questions because anyone may want to add more symbol, e.g. see #11015.
> 
> IMO when one asks new symbol inclusion, he/she should give a justification. For example:
> 
>     * ¼ (one fourth): seems already supported by synthesizers. So give a synthesizer example where this symbol is badly or not announced
> 
>     * all subscript and superscript: give an example of real life where they are used
>       Maybe also the level should be justified.

@feerrenrut could you comment on this please regarding NVAccess point of view? Comments from other people are also welcome.

**Comment by feerrenrut** [comment]
I'm not aware of performance issues with this. I would prefer to address performance problems with the implementation rather than reduce our supported symbols. Even if most synths supply these symbols they may not do so adequately across all languages, or we/a user might not like how they announce the symbol. This gives users the option to use the NVDA supplied symbol name or the rely on the synth. It also lets us adjust symbol names as appropriate, a good example is the Mac command symbol / [looped square](https://en.wikipedia.org/wiki/Looped_square) ⌘ which I believe was added not so long ago. 

I hope this answers your question!

**Comment by CyrilleB79** [comment]
Yes, thanks for your answer.

**Comment by Andre9642** [comment]
> Is there a big performance penalty to add many symbols?

According my tests, the number of symbols is not significant. I tried with a string of 65335 symbols in several locales. However, I don't know if my script is really reliable. There are most likely better ways to measure this.

<details><summary>My script</summary>

```
import globalVars, sys
path = globalVars.appArgs.configPath + "/lib"
if not path in sys.path: sys.path.append(path)

import timeit
import characterProcessing, globalPluginHandler

text = ''.join([chr(c) for c in range (0x20, 0xfffff)])

def test(lang):
	res = characterProcessing.processSpeechSymbols(lang, text, characterProcessing.SYMLVL_SOME)

number = 20
langs = ["ar", "bg", "cs", "da", "fr", "hr", "it", "ja", "ko", "nl", "pl", "ru", "tr", "zh", "zh_TW"]
for lang in langs:
	try:
		res = timeit.timeit(stmt=f"test('{lang}')", number=number, globals=globals())
		b, u = characterProcessing._getSpeechSymbolsForLocale(lang)
		symbolsNumber = len(b.symbols) + len(u.symbols)
		print(f"{lang} -> {symbolsNumber} symbols, total time: {res}, average time: {res/number}")
	except BaseException as err:
		print(f"error with {lang} locale: {err}")
```

</details>

Here's one of my outputs:

```
ar -> 3376 symbols, total time: 221.0103557, average time: 11.050517785
bg -> 3445 symbols, total time: 214.06203370000003, average time: 10.703101685000002
cs -> 3396 symbols, total time: 218.30669819999991, average time: 10.915334909999995
da -> 3445 symbols, total time: 217.11858540000003, average time: 10.85592927
fr -> 3845 symbols, total time: 238.7677773999999, average time: 11.938388869999994
hr -> 3446 symbols, total time: 217.14729369999986, average time: 10.857364684999993
it -> 3446 symbols, total time: 215.80777239999998, average time: 10.790388619999998
ja -> 3968 symbols, total time: 230.4640202999999, average time: 11.523201014999994
ko -> 31709 symbols, total time: 213.701642, average time: 10.685082099999999
nl -> 3441 symbols, total time: 217.8810311999996, average time: 10.89405155999998
pl -> 3448 symbols, total time: 213.60206610000023, average time: 10.680103305000012
ru -> 3493 symbols, total time: 212.6139294999998, average time: 10.63069647499999
tr -> 3672 symbols, total time: 214.91899090000015, average time: 10.745949545000007
zh -> 152 symbols, total time: 214.42888690000018, average time: 10.721444345000009
zh_TW -> 4246 symbols, total time: 211.3798428, average time: 10.56899214
```

> all subscript and superscript: give an example of real life where they are used

In this case, we can remove many symbols :)

**Comment by Adriani90** [comment]
Such symbols are for example often used when writing equations in LateX. Nearly every blind person who uses a laptop or a computer in the school or university works with LateX in Mathematics, Phisics or Chemistry.

Von meinem iPhone gesendet

**Comment by lukaszgo1** [comment]
@Adriani90  wrote:

> Such symbols are for example often used when writing equations in LateX. Nearly every blind person who uses a laptop or a computer in the school or university works with LateX in Mathematics, Phisics or Chemistry. 


The subscripts symbols  are not used in LaTeX as such. When writing in LaTeX the particular notation for them is used, and LaTeX generated pdf's containing math are not accessible enough to be readable.

**Comment by lukaszgo1** [comment]
> * ¼ (one fourth): seems already supported by synthesizers. So give a synthesizer example where this symbol is badly or not announced

With some Polish SAPI 5 synthesizers for these fractions only 
numerator is being read, so having them in NVDA would be beneficial.

**Comment by Adriani90** [comment]
While this applies to pdf probably, there are also other formats you can export latex code to, so I think it is good to have the symbols reported.

 

Von: Łukasz Golonka <notifications@github.com> 
Gesendet: Samstag, 9. Mai 2020 11:08
An: nvaccess/nvda <nvda@noreply.github.com>
Cc: Adriani90 <adriani.botez@googlemail.com>; Mention <mention@noreply.github.com>
Betreff: Re: [nvaccess/nvda] Add symbols from Orca (#11110)

 

@Adriani90 <https://github.com/Adriani90>  wrote:

Such symbols are for example often used when writing equations in LateX. Nearly every blind person who uses a laptop or a computer in the school or university works with LateX in Mathematics, Phisics or Chemistry.

The subscripts symbols are not used in LaTeX as such. When writing in LaTeX the particular notation for them is used, and LaTeX generated pdf's containing math are not accessible enough to be readable.

**Comment by gregjozk** [comment]
Hi,

If I can summarise, most of us agree, that it could be benefitial to
users, if Orca's and such other symbols were included in NVDA?
Personally I vote for inclusion Orcas and other symbols, which has
been used often, but were not included in lcdr.
BTW what's the relationship between NVDA's symbols and included LCDR?

thanks,
Jožef

2020-05-09 11:19 GMT+02.00, Adriani90 <notifications@github.com>:
> While this applies to pdf probably, there are also other formats you can
> export latex code to, so I think it is good to have the symbols reported.

>

>

>

> Von: Łukasz Golonka <notifications@github.com>

> Gesendet: Samstag, 9. Mai 2020 11:08

> An: nvaccess/nvda <nvda@noreply.github.com>

> Cc: Adriani90 <adriani.botez@googlemail.com>; Mention
> <mention@noreply.github.com>

> Betreff: Re: [nvaccess/nvda] Add symbols from Orca (#11110)

>

>

>

**Comment by DrSooom** [comment]
@sthibaul: If you have additional time,, feel free to use my uploaded files regarding [issue #6341](https://github.com/nvaccess/nvda/issues/6341#issuecomment-455053549) as well.

**Comment by feerrenrut** [comment]
I'd like to accept this PR, though I am just waiting on the following [line comment](https://github.com/nvaccess/nvda/pull/11110/files#r421289775) to be addressed: 

> I believe "one quarter" and "three quarters" is much more popular than the also acceptable "fourth" version. I propose this be changed.
> https://english.stackexchange.com/questions/103188/three-quarters-vs-three-fourths

**Comment by sthibaul** [comment]
> I'd like to accept this PR, though I am just waiting on the following [line comment](https://github.com/nvaccess/nvda/pull/11110/files#r421289775) to be addressed:

Ok, fixed so.

File: source/locale/en/symbols.dic [hunk 2]
```diff
@@ -172,6 +230,9 @@ _	line	most
 ⊇	superset of or equal to	none
 
 # Vulgur Fractions U+2150 to U+215E
+¼	one quarter	none
+½	one half	none
+¾	three quarter	none
```
Comment (feerrenrut) [inline source/locale/en/symbols.dic:235]: shouldn't this be "three quarters" (with an 's' on the end)

Comment (sthibaul) [reply source/locale/en/symbols.dic]: "with an s" seems more used indeed

**Review by feerrenrut — APPROVED** [review]
Thanks @sthibaul

File: source/locale/en/symbols.dic [hunk 3]
```diff
@@ -49,7 +49,10 @@ $	dollar	all	norep
 ¢	Cents	all	norep
 ¥	Yen	all	norep
 ₹	Rupee	some	norep
+ƒ	florin	all	norep
+¤	currency sign	all	norep
```
Comment (leonardder) [inline source/locale/en/symbols.dic:53]: I'd say just `currency` is enough

File: source/locale/en/symbols.dic [hunk 4]
```diff
@@ -77,7 +80,16 @@ _	line	most
 {	left brace	most
 }	right brace	most
 |	bar	most
+¦	broken bar	most
 ~	tilda	most
+¡	inverted exclamation point	some
+¿	inverted question mark	some
```
Comment (leonardder) [inline source/locale/en/symbols.dic:86]: Inconsistent with question mark, which we pronounce as `question`

File: source/locale/en/symbols.dic [hunk 5]
```diff
@@ -125,6 +177,11 @@ _	line	most
 ✓	check	some
 ✔	check	some
 🡺	right arrow	some
+†	dagger	some
+‡	double dagger	some
+‣	triangular bullet	none
+✗	x-shaped bullet	none
+￼	object replacement character	none
```
Comment (leonardder) [inline source/locale/en/symbols.dic:184]: Too verbose, see #11177

**Review by LeonarddeR — COMMENTED** [review]
I should have looked at this pr before merge, as I consider some of these symbol pronunciations too verbose and inconsistent to others. Examples below

**Comment by DrSooom** [comment]
@leonardder: Wouldn't it better to add a secondary name field for Unicode characters, which are only spoken in letter navigation? I already planned to create such an issue. This would save us tons of hours of discussions, as then a Unicode character is spoken with its short name in normal navigation and only with its detail name in letter navigation.

**Pull request merged**